test/e2e/util_gatewayapi_test.go

```diff
@@ -701,79 +701,69 @@ type testListener struct {
+			if err := kclient.List(ctx, dnsRecords, client.InNamespace(operatorcontroller.DefaultOperandNamespace)); err != nil {
+				return false, fmt.Errorf("failed to list DNSRecords: %v", err)
 			}
 
-			// Reset the found and foundReady flags for each expectation.
-			found := false
-			foundReady := false
-			// Look for a DNSRecord that matches the expected gateway and DNS name.
-			for _, record := range dnsRecords.Items {
-				if record.Labels["gateway.networking.k8s.io/gateway-name"] == exp.gatewayName &&
-					record.Spec.DNSName == exp.dnsName {
-
-					if !shouldBePresent {
-						expectationsMet = false
-						found = true
-						t.Logf("DNSRecord %q (%s) found but should not be present.", record.Name, exp.dnsName)
-						return false, nil
-					}
-					found = true
-					// DNSRecord found and should be present, check if it is published
-					for _, zone := range record.Status.Zones {
-						for _, condition := range zone.Conditions {
-							if condition.Type == v1.DNSRecordPublishedConditionType && condition.Status == string(metav1.ConditionTrue) {
-								t.Logf("Found DNSRecord %q (%s) %s=%s as expected", record.Name, exp.dnsName, condition.Type, condition.Status)
-								foundReady = true
-							}
-						}
-					}
-					if !foundReady {
-						t.Logf("Found DNSRecord %v but could not determine its readiness; retrying...", record.Name)
-						expectationsMet = false
-						return false, nil
-					}
+			for exp, shouldBePresent := range expectations {
+				expectationMet, err := checkDNSRecordExpectation(t, *dnsRecords, exp, shouldBePresent)
+				if err != nil {
+					return false, err
+				}
+				if !expectationMet {
+					return false, nil
 				}
 			}
 
-			// If the record is expected but not found, return false to continue polling.
-			if shouldBePresent && !found {
-				t.Logf("DNSRecord for hostname %q (gateway: %s) is expected to be present but was not found; retrying...", exp.dnsName, exp.gatewayName)
-				expectationsMet = false
-				return false, nil
-			}
-			// If the record is not expected but was found, return false to continue polling.
-			if !shouldBePresent && found {
-				t.Logf("DNSRecord for hostname %q (gateway: %s) is present but was expected to be absent; retrying...", exp.dnsName, exp.gatewayName)
-				expectationsMet = false
-				return false, nil
-			}
+			return true, nil
+		})
+}
+
+func checkDNSRecordExpectation(t *testing.T, dnsRecords v1.DNSRecordList, exp expectedDnsRecord, shouldBePresent bool) (bool, error) {
+	for _, rec := range dnsRecords.Items {
+		if !dnsRecordMatches(rec, exp) {
+			continue
 		}
-		if haveExpectNotPresent {
-			t.Logf("Continuing polling to ensure non-expected DNSRecords do not exist...")
+		if !shouldBePresent {
+			t.Logf("DNSRecord %s found but expected it absent.", exp.dnsName)
+			return false, nil
 		}
-		return !haveExpectNotPresent, nil
-	})
+		if !isPublished(rec) {
+			t.Logf("DNSRecord %s found but not yet published, retrying...", exp.dnsName)
```

Comment:
See https://github.com/openshift/cluster-ingress-operator/pull/1226/files#r2093564183 on returning this error instead of logging it.